src/Nethermind/Nethermind.Consensus/Processing/BlockCachePreWarmer.cs

```diff
@@ -76,75 +74,144 @@ private void PreWarmCachesParallel(Block suggestedBlock, Hash256 parentStateRoot
+                // If the transaction has already been processed or being processed, exit early
+                if (block.TransactionProcessed > i) return;
+
+                tx = block.Transactions[i];
+                tx.CopyTo(systemTransaction);
+                using IReadOnlyTxProcessingScope scope = env.Build(stateRoot);
+                if (spec.UseTxAccessLists)
+                {
+                    scope.WorldState.WarmUp(tx.AccessList); // eip-2930
+                }
+                TransactionResult result = scope.TransactionProcessor.Trace(systemTransaction, new BlockExecutionContext(block.Header.Clone()), NullTxTracer.Instance);
+                if (_logger.IsTrace) _logger.Trace($"Finished pre-warming cache for tx[{i}] {tx.Hash} with {result}");
+            }
+            catch (Exception ex)
+            {
+                if (_logger.IsDebug) _logger.Error($"Error pre-warming cache {tx?.Hash}", ex);
+            }
+            finally
+            {
+                _systemTransactionPool.Return(systemTransaction);
+                _envPool.Return(env);
+            }
+        });
+    }
+
+    private class AddressWarmer(ParallelOptions parallelOptions, Block block, Hash256 stateRoot, Address? beaconRootsAddress, BlockCachePreWarmer preWarmer)
+        : IThreadPoolWorkItem
+    {
+        private readonly ParallelOptions ParallelOptions = parallelOptions;
+        private readonly Block Block = block;
+        private readonly Hash256 StateRoot = stateRoot;
+        private readonly BlockCachePreWarmer PreWarmer = preWarmer;
+        private readonly Address? BeaconRootsAddress = beaconRootsAddress;
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            IReadOnlyTxProcessorSource env = PreWarmer._envPool.Get();
+            try
+            {
+                using IReadOnlyTxProcessingScope scope = env.Build(StateRoot);
+                WarmupAddresses(ParallelOptions, Block, scope);
+            }
+            catch (Exception ex)
+            {
+                if (PreWarmer._logger.IsDebug) PreWarmer._logger.Error($"Error pre-warming addresses", ex);
+            }
+            finally
+            {
+                PreWarmer._envPool.Return(env);
             }
         }
 
-        void WarmupTransactions(ParallelOptions parallelOptions, IReleaseSpec spec, Block block, Hash256 stateRoot)
+        private void WarmupAddresses(ParallelOptions parallelOptions, Block block, IReadOnlyTxProcessingScope scope)
         {
             if (parallelOptions.CancellationToken.IsCancellationRequested) return;
 
+            // Warm up BeaconRootsAddress first as is first used
+            if (BeaconRootsAddress is not null)
+            {
+                scope.WorldState.WarmUp(BeaconRootsAddress);
+                // Contract accesses this storage cell
+                scope.WorldState.WarmUp(new StorageCell(BeaconRootsAddress, block.Timestamp % 8191));
```

Comment:
StorageCell calculation should be done in `BeaconBlockRootHandler`

Reply:
Moved